### ISSUE TITLE
file_required doesn't work

### ISSUE THREAD

**Issue by 50l3r** [Patroklo/codeigniter3.x-extended-rules#2]
I try to specify this validation lines:

PromocionTitulo:required|max_length[50]|trim
PromocionDescripcion:required|max_length[2000]|trim
PromocionImagen:file_required|file_image_mindim[300,200]|file_allowed_type[image]

with this data:

$_POST
array(2) {
  ["PromocionTitulo"]=>
  string(13) "Promocion API"
  ["PromocionDescripcion"]=>
  string(31) "Descripcion de la promocion api"
}

$_FILES
array(0) {
}

Form Validation passed correctly and should not

**Comment by Patroklo** [comment]
TL;DR : Before, if we made a POST call without issuing a file in the FILE array, even if that file was required it didn't throw an error. That only happened when not adding the input field on the form or sending manually the data without any file. 

Now the library will check if there is a file_required rule and, if there is no file in the FILE array for that field, will add manually the required rule to throw an error from the basic form validation file.

# 

En qué versión de CI? La última?
Oh, y has pasado el post como multipart? Recuerdo que el sistema se volvía loco con el sistema de FILE si no se ponía así

**Comment by 50l3r** [comment]
Si, uso postman 3.0 para ello. Cuando subo una imagen si que me la valida a ver si es del tipo correcto y tiene ciertas dimensiones.

Uso codeigniter 3.0

**Comment by Patroklo** [comment]
Ok, CI 3 ha cambiado bastante internamente desde que cambió de manos, me ha pasado con el router que tengo porque dejaron de funcionar algunas cosas, le echaré un ojo mañana y te comento en cuanto lo haya comprobado.

**Comment by 50l3r** [comment]
Muchas gracias por tu tiempo y tu aporte :+1:

**Comment by Patroklo** [comment]
Hum... una cosa... $_FILES no debería ser empty, debería tener un dato con el campo del fichero con el array de datos vacío y un error de upload tipo id 4 UPLOAD_ERR_NO_FILE.

¿Está el nombre del campo bien definido en el formulario? Igual puede ser eso...

**Comment by 50l3r** [comment]
Estoy desarrollando una api rest, y necesito obligar a subir campos de imágenes. Me explico:

Si yo no informo alguno de los campos PromocionTitulo o PromocionDescripcion, me dice que ese campo falta para poder completar el formulario. En cambio, si no informo el campo PromocionImagen, me da un OK y me ejecuta las acciones oportunas sin tener que subir esa imagen obligatoria.

En el caso de que yo informe PromocionImagen, si que me valida si es una imagen en jpg y con x dimensiones por ejemplo.

**Comment by Patroklo** [comment]
Si, me he fijado que si no se sube una imagen y no se indica que hay un campo de imagen en el formulario, se ejecuta el código normal de validación de formularios. Es un caso raro, ya que se consideraba que siempre iba a existir un campo en el formulario. Como lo que haces es enviar los datos manualmente sin formulario no salta. Le echo un ojo, es de fácil solución por lo que se me ha ocurrido.

**Comment by Patroklo** [comment]
Echale un ojo a ver si ahora funciona. A mí me ha ido.

**Comment by 50l3r** [comment]
Funciona perfecto, muchas gracias por tu aportación.

Tengo por aquí dos funciones más de chequeo que no son muy habituales pero pueden ayudar:

```
    /**
 * [valid_geopos Verificar latitud y longitud correcta]
 * @param  [type] $Coords [Latitud y Longitud]
 * @return [boolean]
 */
public function valid_geopos($Coords){
    if(preg_match('/^[-]?(([0-8]?[0-9])\.(\d+))|(90(\.0+)?);[-]?((((1[0-7][0-9])|([0-9]?[0-9]))\.(\d+))|180(\.0+)?)$/',$Coords)){
        return true;
    }else{
        $this->CI->form_validation->set_message('valid_geopos', 'El mapa %s no tiene una posición correcta');
    }
}

/**
 * [valid_multiple_geopos Verificar latitud y longitud correcta]
 * @param  [type] $Coords [Latitud y Longitud]
 * @return [boolean]
 */
public function valid_multiple_geopos($Coords){
    $json = json_decode($Coords);

    if(!empty($json)){
        foreach($json as $pos){
            if(!preg_match('/^[-]?(([0-8]?[0-9])\.(\d+))|(90(\.0+)?);[-]?((((1[0-7][0-9])|([0-9]?[0-9]))\.(\d+))|180(\.0+)?)$/',$Coords)){
                $this->CI->form_validation->set_message('valid_multiple_geopos', 'El mapa de límites %s no tiene una posición correcta');
            }
        }

        return true;
    }

    $this->CI->form_validation->set_message('valid_multiple_geopos', 'El mapa de límites %s no tiene una posición correcta');
}
```

Sirven para validar latitudes y longitudes de manera independiente o multiples recogidas mediante json.

**Comment by Patroklo** [comment]
Mola, los añadiré en un rato. Gracias.
